build_runner/lib/src/watcher/graph_watcher.dart

```diff
@@ -38,57 +39,34 @@ class PackageGraphWatcher {
 
   /// Returns a stream of records for assets that changed in the package graph.
   Stream<AssetChange> watch() {
-    if (controller != null) return controller.stream;
-    List<StreamSubscription> subscriptions;
-    controller = StreamController<AssetChange>(
-      sync: true,
-      onListen: () {
-        subscriptions = logTimedSync(
-          _logger,
-          'Setting up file watchers',
-          () => _watch(controller),
-        );
-      },
-      onCancel: () {
-        for (final subscription in subscriptions) {
-          subscription.cancel();
-        }
-        _readyCompleter = Completer<Null>();
-        var done = controller.close();
-        controller = null;
-        return done;
-      },
-    );
-    return controller.stream;
+    assert(!_isWatching);
+    _isWatching = true;
+    return LazyStream(
+        () => logTimedSync(_logger, 'Setting up file watchers', _watch));
   }
 
-  List<StreamSubscription> _watch(StreamSink<AssetChange> sink) {
-    final subscriptions = <StreamSubscription>[];
-    var allWatchers = <PackageNodeWatcher>[];
-    _graph.allPackages.forEach((name, node) {
-      if (node.dependencyType == DependencyType.hosted ||
-          node.dependencyType == DependencyType.github) {
-        return;
-      }
-      final nestedPackages = _nestedPaths(node);
-      final nodeWatcher = _strategy(node);
-      allWatchers.add(nodeWatcher);
-      subscriptions.add(nodeWatcher.watch().listen((event) {
-        // TODO: Consider a faster filtering strategy.
-        if (nestedPackages.any((path) => event.id.path.startsWith(path))) {
-          return;
-        }
-        sink.add(event);
-      }));
-    });
+  Stream<AssetChange> _watch() {
+    final allWatchers = _graph.allPackages.values
+        .where((node) => !(node.dependencyType == DependencyType.hosted ||
```

Comment:
Not directly related, but we may want to consider changing this to just `node.dependencyType == DependencyType.path`? For instance the there are more dependency types, like SDK (for the flutter package).

Reply:
Good idea